File: typespec/common/openings.ts
export type OpeningState =
  | "DRAFT_OPENING_STATE"
  | "ACTIVE_OPENING_STATE"
  | "SUSPENDED_OPENING_STATE"
  | "CLOSED_OPENING_STATE";

export const OpeningStates = {
  DRAFT: "DRAFT_OPENING_STATE" as OpeningState,
  ACTIVE: "ACTIVE_OPENING_STATE" as OpeningState,
  SUSPENDED: "SUSPENDED_OPENING_STATE" as OpeningState,
  CLOSED: "CLOSED_OPENING_STATE" as OpeningState,
} as const;

export type OpeningType =
  | "FULL_TIME_OPENING"
  | "PART_TIME_OPENING"
  | "CONTRACT_OPENING"
  | "INTERNSHIP_OPENING"
  | "UNSPECIFIED_OPENING";

export const OpeningTypes = {
  FULL_TIME: "FULL_TIME_OPENING" as OpeningType,
  PART_TIME: "PART_TIME_OPENING" as OpeningType,
  CONTRACT: "CONTRACT_OPENING" as OpeningType,
  INTERNSHIP: "INTERNSHIP_OPENING" as OpeningType,
  UNSPECIFIED: "UNSPECIFIED_OPENING" as OpeningType,
} as const;

export type EducationLevel =
  | "BACHELOR_EDUCATION"
  | "MASTER_EDUCATION"
  | "DOCTORATE_EDUCATION"
  | "NOT_MATTERS_EDUCATION"
  | "UNSPECIFIED_EDUCATION";

export const EducationLevels = {
  BACHELOR: "BACHELOR_EDUCATION" as EducationLevel,
  MASTER: "MASTER_EDUCATION" as EducationLevel,
  DOCTORATE: "DOCTORATE_EDUCATION" as EducationLevel,
  NOT_MATTERS: "NOT_MATTERS_EDUCATION" as EducationLevel,
  UNSPECIFIED: "UNSPECIFIED_EDUCATION" as EducationLevel,
} as const;

export function isValidOpeningState(state: string): state is OpeningState {
  return Object.values(OpeningStates).includes(state as OpeningState);
}

export function isValidOpeningType(type: string): type is OpeningType {
  return Object.values(OpeningTypes).includes(type as OpeningType);
}

export function isValidEducationLevel(level: string): level is EducationLevel {
  return Object.values(EducationLevels).includes(level as EducationLevel);
}

File: typespec/common/openings.go
package common

type OpeningState string

const (
	DraftOpening     OpeningState = "DRAFT_OPENING_STATE"
	ActiveOpening    OpeningState = "ACTIVE_OPENING_STATE"
	SuspendedOpening OpeningState = "SUSPENDED_OPENING_STATE"
	ClosedOpening    OpeningState = "CLOSED_OPENING_STATE"
)

type OpeningType string

const (
	// Any changes to here should be reflected also in the IsValid() method
	FullTimeOpening    OpeningType = "FULL_TIME_OPENING"
	PartTimeOpening    OpeningType = "PART_TIME_OPENING"
	ContractOpening    OpeningType = "CONTRACT_OPENING"
	InternshipOpening  OpeningType = "INTERNSHIP_OPENING"
	UnspecifiedOpening OpeningType = "UNSPECIFIED_OPENING"
)

func (o OpeningType) IsValid() bool {
	switch o {
	case FullTimeOpening,
		PartTimeOpening,
		ContractOpening,
		InternshipOpening,
		UnspecifiedOpening:
		return true
	}
	return false
}

type EducationLevel string

const (
	BachelorEducation    EducationLevel = "BACHELOR_EDUCATION"
	MasterEducation      EducationLevel = "MASTER_EDUCATION"
	DoctorateEducation   EducationLevel = "DOCTORATE_EDUCATION"
	NotMattersEducation  EducationLevel = "NOT_MATTERS_EDUCATION"
	UnspecifiedEducation EducationLevel = "UNSPECIFIED_EDUCATION"
)

func (e EducationLevel) IsValid() bool {
	return e == BachelorEducation || e == MasterEducation ||
		e == DoctorateEducation ||
		e == NotMattersEducation ||
		e == UnspecifiedEducation
}

type Salary struct {
	MinAmount float64  `json:"min_amount" validate:"required,min=0"`
	MaxAmount float64  `json:"max_amount" validate:"required,min=1"`
	Currency  Currency `json:"currency"   validate:"required"`
}

type FilterOpeningTagsRequest struct {
	Prefix *string `json:"prefix,omitempty"`
}

type OpeningTagID string

type OpeningTag struct {
	ID   OpeningTagID `json:"id"   validate:"required"`
	Name string       `json:"name" validate:"required"`
}

type OpeningTags struct {
	Tags []OpeningTag `json:"tags" validate:"required"`
}
